README.md
[![Build Status](https://travis-ci.org/kmadisa/mind-your-stonks.svg?branch=master)](https://travis-ci.org/kmadisa/mind-your-stonks)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
![GitHub last commit](https://img.shields.io/github/last-commit/kmadisa/mind-your-stonks.svg)
![GitHub issues](https://img.shields.io/github/issues/kmadisa/mind-your-stonks.svg)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/ce9da614c0034a3ca373513e119705e0)](https://www.codacy.com/manual/katleho.madisa47/mind-your-stonks?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=kmadisa/mind-your-stonks&amp;utm_campaign=Badge_Grade)

# Mind your *stonks*


![Screen_Shot_2019-06-05_at_1 26 32_PM](https://user-images.githubusercontent.com/16665803/61865197-15e25e00-aed3-11e9-8541-4fff382916b7.jpg)

## The Story
[BET.co.za](https://bet.co.za) does not have an easier way to track the changes in the user's account balance. All that it displays is the current account balance. In order to get an idea of **how much you had vs how much you have** in your account, you have to sift through lots and lots of pages on the transaction history, not to mention the brutal experience the user will have to go through to compute the **account's balance at a particular moment in time** from all that information.

The purpose of this project is to provide [BET.co.za](https://bet.co.za) clients an easier way to keep track of the changes occurring in their account balance.

Basically there is utility script that run and log into the client's [BET.co.za](https://bet.co.za) account (using [Selenium](https://selenium-python.readthedocs.io/)) and read the value of the current balance and uploads the data to a [Google Sheets](https://docs.google.com/spreadsheets/u/0/), where then it can be visualized.


&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; **AND JUST LIKE THAT YOU KNOW THE RISE & FALL OF YOUR STONKS !!!!**


## Get Started

1. Obtain Google API for authentication:
    *   Follow the instructions [here](https://gspread.readthedocs.io/en/latest/oauth2.html#oauth-credentials)

2. Install ensure that `geckodrive` for Firefox is installed.
    *   Download [geckodriver](https://github.com/mozilla/geckodriver)
        *   ```wget https://github.com/mozilla/geckodriver/releases/download/v0.24.0/geckodriver-v0.24.0-linux64.tar.gz```
        *   Extract: ```tar -xvzf geckodriver-v0.24.0-linux64.tar.gz```
    *   `sudo cp geckodriver /usr/local/bin`

3. Install Selenium:
    *   `pip install selenium`

4. Upload a copy of the [spreadsheet](https://docs.google.com/spreadsheets/d/1k--fJt5qC191RMHH3D2MbhRhaIJb__WTEBjOL1rcksc/edit?usp=sharing) to your own GDrive or [GSpeadsheet](https://docs.google.com/spreadsheets).

![Screenshot from 2019-12-10 15-32-48](https://user-images.githubusercontent.com/16665803/70533841-8debb080-1b62-11ea-82a6-a4aa9e188ef3.png)
Figure 1. A snapshot of the spreadsheet and the respective columns (Note: Numbers may or may not have been doctored!).

#### Table columns
   * *Date*: date reading was made (yyyy-mm-dd).
   * *Timestamp*: indicated when the script logged into the account hh:mm PM/AM.
   * *Balance*: the current balance in the account in S.A rands.
   * *Money in bets*: the amount of money placed in a bet which is still unresolved.
   * *Actual Loss/Gain*: difference between the previous known balance and the current one
                         (+ money in bets).
   * *% Increase*: calculated from the previous known balance and the current one (+ money in bets).

##### Other fields
   * *Opening Balance*: te amount of money in the account at the first of the month.
   * *Closing Balance*: te amount of money in the account at the end of the current month.
   * *Graph*: a plot of the account's balance.

## Usage

```bash
python query_balance.py -h
usage: query_balance.py [-h] [--update-spreadsheet UPDATE_SPREADSHEET]
                        username password

Scrape the BET.co.za website to obtain the account balance. It also writes the
data to a Google Spreadsheet.

positional arguments:
  username              Bet.co.za registered email address
  password              Bet.co.za account password.

optional arguments:
  -h, --help            show this help message and exit
  --update-spreadsheet UPDATE_SPREADSHEET
                        Update spreadsheet with new data. This requires the
                        client_secret.json file for authentication. It is
                        downloaded from the Google Developers' Console.
```

Typical usage:
```bash
query_balance.py $USERNAME $PASSWORD --update-spreadsheet ./client_secret.json
```

## Travis CI automated daily balance reader
Travis CI can automatically run your Google App Engine based application, by encrypting your `clients_secrets.json` file and pushing it to `GitHub`.
See: https://docs.travis-ci.com/user/deployment/google-app-engine/


## Feedback

Feel free to fork it or send me a PR to improve it.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

### Removed

### Changed

## [1.0.1]

### Changed

- Setting the ```| Loss/Gain | % Decrease/Increase |``` values to zero on the first day of the month [PR#30](https://github.com/kmadisa/mind-your-stonks/pull/30).

## [1.0.0] 2019-12-10

### Added

- Created a `BetClient` class encapsulating all the behaviour in the deleted `bet_stats.py` and `spreadsheet.py` modules [PR#15](https://github.com/kmadisa/mind-your-stonks/pull/15). The class is defined in the `bet_client.py` module.
- A utility classs that wraps the Selenium webdriver object [PR#15](https://github.com/kmadisa/mind-your-stonks/pull/15).

### Removed

- Deleted the `bet_stats.py` and `spreadsheet.py` scripts [PR#15](https://github.com/kmadisa/mind-your-stonks/pull/15).
- Usage of the [python-gspread-sheets](https://github.com/CROSP/python-gspread-sheets) package [PR#27](https://github.com/kmadisa/mind-your-stonks/pull/27).

### Changed

- Support for the new spreadsheet with nicely formatted tables for easy use.
- Moved around the spreadsheet columns:
    ```| Date | Timestamp | Balance | Money in bets | Loss/Gain | % Decrease/Increase |```
    [PR#27](https://github.com/kmadisa/mind-your-stonks/pull/27).

## [0.0.1] - 2019-10-26

### Added

- This CHANGELOG file to hopefully serve as an evolving example of a
  standardized open source project CHANGELOG.
- README has instructions on how to use the project [PR#14](https://github.com/kmadisa/mind-your-stonks/pull/14).
- Added support to be able to navigate through pages [PR#12](https://github.com/kmadisa/mind-your-stonks/pull/1)2.
- Computing the amount of money that has been placed in bets [PR#9](https://github.com/kmadisa/mind-your-stonks/pull/9).
- Added scripts to scrape [BET.co.za](https://www.bet.co.za) site and writes the data
  to the Google sheet [PR#1](https://github.com/kmadisa/mind-your-stonks/pull/1).


sandbox/test_script.py
#!/usr/bin/env python3
from selenium import webdriver
from selenium.webdriver.firefox.options import Options
from mind_your_stonks.bet_client import BetClient

bc = BetClient("k", "*")
options = Options()
options.headless = True

drv = webdriver.Firefox(options=options)
drv.get('file:///home/kmadisa47/Desktop/ticket3.html')
bc.driver = drv

bet_data = bc._extract_ticket_data()

for metadata, fields, table in zip(bet_data[0], bet_data[1], bet_data[2]):
    print(metadata)
    print(fields)
    print("###########################")
    for t in table:
        print(t)
    print("###########################")


tests/test_bet_client.py
import pytest
from mock import Mock, call, patch

from selenium import webdriver
from selenium.webdriver.firefox.options import Options
from selenium.webdriver.firefox.webelement import FirefoxWebElement
from  selenium.common.exceptions import TimeoutException
from mind_your_stonks.bet_client import BetClient, BET_URL, TIMEOUT


@pytest.fixture(scope="class")
def bet_client():
    _username = "pytester"
    _password = "pytester3.6.8"
    bet_client = BetClient(_username, _password)

    yield bet_client

@pytest.fixture()
def driver():
    options = Options()
    options.headless = True
    driver = webdriver.Firefox(options=options)
    driver.get(
        "file:///home/kmadisa47/src/personal/mind-your-stonks/sandbox/sign_in.html")

    yield driver

@pytest.mark.usefixtures("bet_client")
class TestBetClient(object):

    def test_sign_in(self, bet_client, driver):

        bet_client.driver = driver

        bet_client.web_setup.open_session = Mock()

        FirefoxWebElement.send_keys = Mock()
        FirefoxWebElement.click = Mock()

        with pytest.raises(TimeoutException):
            bet_client.sign_in()

        assert FirefoxWebElement.send_keys.call_count == 2
        assert FirefoxWebElement.send_keys.call_args_list == [
            call("pytester"), call("pytester3.6.8")
        ]
        assert FirefoxWebElement.click.call_count == 2

    def test_sign_out(self, bet_client, driver):
        pass

    def test_timestamp(self, bet_client):
        pass

    def test_goto_betting_history(self):
        pass

    def test_goto_account_history(self):
        pass

    def test_filter_betting_history(self):
        pass

    def test_export_betting_history_data(self):
        pass


scripts/bet_stats.py
import argparse
import time
import csv

from datetime import datetime

from selenium import webdriver 
from selenium.webdriver.firefox.options import Options
from selenium.webdriver.support.select import Select 

(TICKET, EVENT_DATE, TOURNAMENT, EVENT, SELECTION, BET_TYPE,
 STAKE, POTENTIAL_WIN, STATUS)  = range(1, 10)

parser = argparse.ArgumentParser(description="Scrape the BET.co.za website"
                                 " to obtain the account balance.")
parser.add_argument(
    "--username",
    required=True,
    help="Bet.co.za registered email address",)
parser.add_argument(
    "--password",
    required=True, 
    help="Bet.co.za account password.",)

def main():

    opts = parser.parse_args()
    # How to get the account balance.
    options = Options()
    options.headless = True
    driver = webdriver.Firefox(options=options)
    driver.get("https://www.bet.co.za")

    driver.find_element_by_name("frmUsername").send_keys(opts.username)
    driver.find_element_by_name("frmPassword").send_keys(opts.password)
    driver.find_element_by_name("frmForceTerms").click()
    driver.find_element_by_name("submitted").click()
    
    timestamp = driver.find_element_by_id("time").text.split("Your time: ")
    timestamp = timestamp[-1].strip()
    account_balance = driver.find_element_by_id("blocklogout_userBalanceText").text
    date = datetime.date(datetime.now())


    # Also need to get the amount of funds that is placed in bets
    driver.find_element_by_link_text('My Betting History').click()
    # Get the filter form
    form_filter = driver.find_element_by_id('filter_form')
    # Create a selector object for dropdown tables
    selector = Select(form_filter.find_element_by_id('status'))
    # Select 'Unsettled' bets
    selector.select_by_visible_text('Unsettled')
    # Click on the 'Go' button to filter bets
    form_filter.find_element_by_class_name('inputBtn').click()
    # Get the table object
    table = driver.find_element_by_class_name('stdTable')

    # Pages can come in different forms
    # '' ==> means just one page
    # '12»' ==> means just two pages in total
    # '1234567»[12]' ==> means 12 pages in total
    # 
    pagination = driver.find_element_by_class_name("pagination")
    pagination_text = pagination.text
    num_of_pages = 0
    if pagination_text == '':
        # only one page to deal with
        num_of_pages = 1
    elif pagination_text.endswith("»"):
        # '12»' -> 2
        num_of_pages = int(pagination_text.split('»')[0][-1])
    elif pagination_text.endswith("]"):
        # '1234567»[12]' -> 12
        num_of_pages = int(pagination_text.split('[')[-1].split(']')[0])

    money_in_bets = 0.00

    for page in range(1, num_of_pages+1):
        if page > 1:
            pagination = driver.find_element_by_class_name("pagination") # Need of raises StaleElementReferenceException
            page_ = pagination.find_element_by_link_text('{}'.format(page))
            page_.click()

        # Get all the rows on column number 7 (Stake)
        stakes = table.find_elements_by_xpath("//tr/td["+str(STAKE)+"]")
        
        for stake in stakes:
            money_in_bets += float(stake.text)

    driver.close()

    with open('balance.csv', mode='w') as balance_file:
        balance_writer = csv.writer(balance_file, delimiter=',',
                                    quotechar='"', quoting=csv.QUOTE_MINIMAL)
        balance_writer.writerow([date, account_balance, timestamp, money_in_bets])

if __name__ == "__main__":
    main()


scripts/spreadsheet.py
import argparse
import csv

import gspread
from oauth2client.service_account import ServiceAccountCredentials

parser = argparse.ArgumentParser(description="Populate the Google spreadsheet"
                                 " with data from bet.co.za")
parser.add_argument(
    "--client-secrets",
    required=True,
    help="The client_secret.json file downloaded from the Google Developers'"
        " Console.",)

def main():

    opts = parser.parse_args()
    # use creds to create a client to interact with the Google Drive API
    scope = ['https://spreadsheets.google.com/feeds',
             'https://www.googleapis.com/auth/drive']
    creds = ServiceAccountCredentials.from_json_keyfile_name(opts.client_secrets, scope)
    client = gspread.authorize(creds)

    # Find a workbook by name and open the first sheet
    # Make sure you use the right name here.
    sheet = client.open("BET account balance").sheet1

    with open('balance.csv', mode='r') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        for row in csv_reader:
            previous_row_num = len(sheet.get_all_values())
            current_row_num = previous_row_num + 1
            percentage_increase = "=ROUND(MINUS(B{},B{})/B{}, 2)".format(
                current_row_num, previous_row_num, previous_row_num)
            row.insert(2, percentage_increase)  # Move the timestamp to the 2nd last column
                                                 # to preserve the structure of the table.
            sheet.append_row(row, value_input_option='USER_ENTERED')

if __name__ == "__main__":
    main()


mind_your_stonks/bet_client.py
from aenum import Constant
from datetime import datetime

from selenium.webdriver.support.select import Select

from mind_your_stonks.web_driver import WebDriverSetup


BET_URL = "https://www.bet.co.za"

# Solution for creating a set of constants found here
# https://codereview.stackexchange.com/questions/193090/python-constant-class-different-enum-implementation

# Betting History table columns
# | Ticket | Event Date | Tournament | Event | Selection | Bet Type | Stake | Potential Win | Status |


class BetHistoryTableColumn(Constant):
    (TICKET, EVENT_DATE, TOURNAMENT, EVENT, SELECTION, BET_TYPE,
     STAKE, POTENTIAL_WIN, STATUS) = range(1, 10)


class BetStatus(Constant):
    ALL = "Filter by wager status"
    UNSETTLED = "Unsettled"
    WON = "Won"
    LOST = "Lost"
    VOID = "Void"


class BetMonth(Constant):
    (JANUARY, FEBRUARY, MARCH, APRIL, MAY, JUNE, JULY,
     AUGUST, SEPTEMBER, OCTOBER, NOVEMBER, DECEMBER) = range(1, 13)
    LAST_7_DAYS = "7d"
    LAST_30_DAYS = "30"
    ALL = "all"


class BetClient(object):
    """The BETcoza client allows users to navigate around the https://www.bet.co.za
    site and extract data from it.

    Attributes
    ----------
    driver : webdriver.Firefox instance
        Controls a browser by sending commands to a remote server.
    """
    def __init__(self, username, password, headless=True):
        """
        Parameters
        ----------
        username : str
            Registered username for the https://www.bet.co.za site.
        password : str
            Registered password for the https://www.bet.co.za site.
        headless : bool
            Flag to run the Firefox browser headless or not.
        """
        self._username = username
        self._password = password
        self.web_setup = WebDriverSetup(headless)
        self.driver = self.web_setup.driver

    def sign_in(self):
        """Log into the https://www.bet.co.za site using your credentials.
        """
        self.web_setup.open_session(BET_URL)
        self.driver.find_element_by_name("frmUsername").send_keys(self._username)
        self.driver.find_element_by_name("frmPassword").send_keys(self._password)
        self.driver.find_element_by_name("frmForceTerms").click()
        self.driver.find_element_by_name("submitted").click()

    def sign_out(self):
        """Log out of the https://www.bet.co.za site.
        """
        self.driver.find_element_by_xpath("//*[@id='block-logout']").click()
        self.web_setup.close_session()

    @property
    def timestamp(self):
        """Returns the current time displayed on https://www.bet.co.za site.

        Returns
        -------
        timestamp : str
            Current timestamp displayed on the site.
        """
        timestamp = self.driver.find_element_by_id("time").text.split("Your time: ")
        timestamp = timestamp[-1].strip()
        return timestamp

    @property
    def current_balance(self):
        """Returns the current account balance displayed on https://www.bet.co.za site.
        Returns
        -------
        account_balance : str
            Current account balance displayed on the site.
        """
        account_balance = (
            self.driver.find_element_by_id("blocklogout_userBalanceText").text)
        return account_balance

    def goto_betting_history(self):
        """Navigate the https://www.bet.co.za betting history page.
        """
        self.driver.find_element_by_link_text("My Betting History").click()

    def goto_account_history(self):
        """Navigate the https://www.bet.co.za account history page.
        """
        self.driver.find_element_by_link_text("My Account History").click()

    def filter_betting_history(self, status, month=BetMonth.LAST_7_DAYS,
                               year=str(datetime.now().year)):
        """Filter the https://www.bet.co.za betting history table according to the
        filter options.

        Parameters
        ----------
        status : BetStatus instance.
            Wager status.
        month : BetMonth instance
            Date filter value.
        year : str
            Years from 2011 to current.
        """
        # Get the filter form
        form_filter = self.driver.find_element_by_id("filter_form")
        # Create a selector object for dropdown tables
        selector = Select(form_filter.find_element_by_id("status"))
        # Select option in wager status dropdown
        selector.select_by_visible_text(status)

        if month != BetMonth.LAST_7_DAYS:
            selector = Select(form_filter.find_element_by_class_name("date_range"))
            # Select option in month dropdown
            selector.select_by_visible_text(month)

        if year != str(datetime.now().year):
            selector = Select(form_filter.find_element_by_class_name("year"))
            # Select option in year dropdown
            selector.select_by_visible_text(year)

        # Click on the 'Go' button to filter bets
        form_filter.find_element_by_class_name("inputBtn").click()

    def _get_number_of_pages_for_table(self):
        """

        Returns
        -------
        num_of_pages : int
            The maximum number of pages for the table.
        """
        # Pages can come in different forms
        # '' ==> means just one page
        # '12»' ==> means just two pages in total
        # '1234567»[12]' ==> means 12 pages in total
        pagination = self.driver.find_element_by_class_name("pagination")
        pagination_text = pagination.text
        num_of_pages = 0
        if pagination_text == "":
            # only one page to deal with
            num_of_pages = 1
        elif pagination_text.endswith("»"):
            # '12»' -> 2
            num_of_pages = int(pagination_text.split('»')[0][-1])
        elif pagination_text.endswith("]"):
            # '1234567»[12]' -> 12
            num_of_pages = int(pagination_text.split("[")[-1].split("]")[0])

        return num_of_pages

    def compute_money_invested(self):
        """Calculate the amount of money has been taken from the balance and placed
        in a bet(s).

        Returns
        -------
        money_invested : float
            The amount of money placed in a bet(s).
        """
        money_invested = 0.00
        num_of_pages = self._get_number_of_pages_for_table()
        # Get the table object
        table = self.driver.find_element_by_class_name("stdTable")
        for page in range(1, num_of_pages+1):
            if page > 1:
                # Need to get the pagination element again or else raises
                # StaleElementReferenceException
                pagination = self.driver.find_element_by_class_name("pagination")
                page_ = pagination.find_element_by_link_text('{}'.format(page))
                page_.click()

            # Get all the rows on column number 7 (Stake)
            stakes = table.find_elements_by_xpath(
                "//tr/td["+str(BetHistoryTableColumn.STAKE)+"]")

            for stake in stakes:
                money_invested += float(stake.text)

        return money_invested


mind_your_stonks/web_driver.py
import time
import random
import psutil

from selenium import webdriver
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.firefox.options import Options
from loguru import logger

TIMEOUT = 60


def sleeper(min_range=10, max_range=30):
    time.sleep(random.randint(min_range, max_range))


class WebDriverSetup(object):

    def __init__(self, headless):
        self._timeout = TIMEOUT
        self._options = Options()
        self._options.headless = headless
        self._profile = self._disable_images_firefox_profile()
        self.logger = logger

        self.driver = webdriver.Firefox(
            firefox_profile=self._profile, options=self._options, timeout=self._timeout
        )
        time.sleep(0.3)
        if self._options.headless:
            self.logger.info("Headless Firefox Initialized.")

    def open_session(self, url):
        try:
            self.logger.debug(f"Navigating to {url}.")
            self.driver.get(url)
            self.driver.set_page_load_timeout(self._timeout)
            self.logger.debug("Successfully opened the url.")
            time.sleep(0.5)
        except TimeoutException:
            self.logger.exception("Timed-out while loading page.")
            self.close_session()

    def _disable_images_firefox_profile(self):
        """Summary
        Returns:
            Object: FirefoxProfile
        """
        # get the Firefox profile object
        firefoxProfile = webdriver.FirefoxProfile()
        # Disable images
        firefoxProfile.set_preference("permissions.default.image", 2)
        # Disable Flash
        firefoxProfile.set_preference(
            "dom.ipc.plugins.enabled.libflashplayer.so", "false"
        )
        # Set the modified profile while creating the browser object
        return firefoxProfile

    def close_session(self):
        """Close browser and cleanup"""
        self.logger.info("Closing the browser...")
        self.driver.close()
        self.driver.quit()
        sleeper()
        PROCNAME = "geckodriver"
        self.logger.info("Cleaning up by killing {} process", PROCNAME)
        _ = [
            proc.terminate()
            for proc in psutil.process_iter()
            if proc.name() == PROCNAME
        ]
        self.logger.info("Done...")


scripts/query_betting_history.py
#!/usr/bin/env python3
import csv
import argparse
from datetime import datetime

from aenum import Constant
from loguru import logger

import gspread
from oauth2client.service_account import ServiceAccountCredentials

from googlesheets.google_sheet_manager import GoogleSheetManager

from mind_your_stonks.bet_client import BetClient, BetStatus



parser = argparse.ArgumentParser(description="Scrape the BET.co.za website"
                                 " to obtain the account balance. It also"
                                 " writes the data to a Google Spreadsheet.")
parser.add_argument(
    "username",
    help="Bet.co.za registered email address",)
parser.add_argument(
    "password",
    help="Bet.co.za account password.",)
parser.add_argument(
    "--update-spreadsheet",
    help="Update spreadsheet with new data. This requires the client_secret.json"
        " file for authentication. It is downloaded from the Google Developers' Console.",)

def main():

    opts = parser.parse_args()

    better = BetClient(opts.username, opts.password)
    better.sign_in()
    
    better.goto_betting_history()
    bets = better.export_betting_history_data()

    better.sign_out()

    with open('betting_history_archive.csv', mode='w') as archive:
            archive_writer = csv.writer(archive, delimiter=" ", quotechar='"', quoting=csv.QUOTE_ALL)
            for bet in bets:
                archive_writer.writerow(bet)

    if opts.update_spreadsheet:

        gsheet_manager = GoogleSheetManager(credentials_path=opts.update_spreadsheet,
                                            sheet_id="1DWRxtnTIHiRuQ9w1Yq4TiZJqidIZhguiGadYDNgnZsI")
        gsheet_manager.start_session()

        previous_row_num = len(gsheet_manager.get_all_rows())
        current_row_num = previous_row_num + 1

        # Write to the spreadsheet
        # Spreadsheet columns
        # | Date | Timestamp | Money in bets | Balance | Loss/Gain | % Increase |

        gsheet_manager.close_session()

if __name__ == "__main__":
    main()


scripts/query_balance.py
#!/usr/bin/env python3
import argparse
from datetime import datetime
from aenum import Constant
from loguru import logger

import gspread
from oauth2client.service_account import ServiceAccountCredentials

from mind_your_stonks.bet_client import BetClient, BetStatus


class SpreadsheetColumn(Constant):
    (DATE, TIMESTAMP, BALANCE, MONEY_IN_BETS,
     GAIN_LOSS, PERCENTAGE_INCREASE) = map(chr, range(65, 71))


OPENING_BALANCE_ROW = 4
CLOSING_BALANCE_ROW = 5


parser = argparse.ArgumentParser(description="Scrape the BET.co.za website"
                                 " to obtain the account balance. It also"
                                 " writes the data to a Google Spreadsheet.")
parser.add_argument(
    "username",
    help="Bet.co.za registered email address",)
parser.add_argument(
    "password",
    help="Bet.co.za account password.",)
parser.add_argument(
    "--update-spreadsheet",
    help="Update spreadsheet with new data. This requires the client_secret.json"
         " file for authentication. It is downloaded from the Google Developers'"
         " Console.",)


def is_leap_year(year):
    # Python program to check if year is a leap year or not
    if (year % 4) == 0:
        if (year % 100) == 0:
            if (year % 400) == 0:
                return True
            else:
                return False
        else:
            return True
    else:
        return False


def main():

    opts = parser.parse_args()

    better = BetClient(opts.username, opts.password)
    better.sign_in()

    table_entry = {}
    table_entry["Date"] = "=DATE({},{},{})".format(
        *(str(datetime.date(datetime.now())).split("-")))
    table_entry["Timestamp"] = better.timestamp

    better.goto_betting_history()
    better.filter_betting_history(BetStatus.UNSETTLED)

    table_entry["Balance"] = better.current_balance
    table_entry["Money_in_bets"] = better.compute_money_invested()

    better.sign_out()

    if opts.update_spreadsheet:

        date_today = datetime.date(datetime.now())
        day = date_today.day
        month = date_today.strftime("%B")  # Get the month as text
        year = date_today.year
        months_with_30_days = ["April", "June", "September", "November"]
        leap_year_month = "February"

        scope = ['https://spreadsheets.google.com/feeds',
                 'https://www.googleapis.com/auth/drive']
        # Use creds to create a client to interact with the Google Drive API
        creds = ServiceAccountCredentials.from_json_keyfile_name(
            opts.update_spreadsheet, scope)
        client = gspread.authorize(creds)
        # Find a workbook by name and open the sheet of interest.
        # Make sure you use the right name here.
        spreadsheet = client.open("BET account balance")
        logger.debug("Started session on the Google sheet.")
        sheet = spreadsheet.worksheet("{} {}".format(month, year))

        # Update the opening/closing balance
        current_account_balance = table_entry["Balance"]
        if day == 1:
            sheet.update_cell(
                OPENING_BALANCE_ROW, SpreadsheetColumn.BALANCE, current_account_balance)
        elif day == 28 and month == leap_year_month and not is_leap_year(year):
            sheet.update_cell(
                CLOSING_BALANCE_ROW, SpreadsheetColumn.BALANCE, current_account_balance)
        elif day == 29 and month == leap_year_month:
            sheet.update_cell(
                CLOSING_BALANCE_ROW, SpreadsheetColumn.BALANCE, current_account_balance)
        elif day == 30 and month in months_with_30_days:
            sheet.update_cell(
                CLOSING_BALANCE_ROW, SpreadsheetColumn.BALANCE, current_account_balance)
        elif day == 31:
            sheet.update_cell(
                CLOSING_BALANCE_ROW, SpreadsheetColumn.BALANCE, current_account_balance)

        previous_row_num = len(sheet.get_all_values())
        current_row_num = previous_row_num + 1

        if day == 1:
            table_entry["Gain_loss"] = 0.00
            table_entry["Percentage_increase"] = 0.00
        else:
            table_entry["Gain_loss"] = (
                "=MINUS(SUM({0}{2},{1}{2}), SUM({0}{3},{1}{3}))".format(
                    SpreadsheetColumn.BALANCE, SpreadsheetColumn.MONEY_IN_BETS,
                    current_row_num, previous_row_num)
            )

            table_entry["Percentage_increase"] = (
                "={2}{3}/SUM({0}{4}, {1}{4})".format(
                    SpreadsheetColumn.BALANCE, SpreadsheetColumn.MONEY_IN_BETS,
                    SpreadsheetColumn.GAIN_LOSS, current_row_num, previous_row_num)
            )
        # Write to the spreadsheet
        # Spreadsheet columns
        # | Date | Timestamp | Balance | Money in bets | Actual Loss/Gain | % Decrease/Increase |
        logger.debug("COMMENCING: Writing data to the Google sheet.")
        sheet.append_row(list(table_entry.values()), value_input_option='USER_ENTERED')
        logger.debug("COMPLETE: Writing data to the Google sheet.")


if __name__ == "__main__":
    main()
